src/mlpack/methods/ann/reduction_rules/add_reduction.hpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * @file methods/ann/reduction_rules/add_reduction.hpp
+ * @author Shubham Agrawal
+ *
+ * Reduction of a given network with add reduction rule.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_REDUCTION_RULES_ADD_REDUCTION_HPP
+#define MLPACK_METHODS_ANN_REDUCTION_RULES_ADD_REDUCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * This class is used to reduce the network with the add reduction rule.
+ * @tparam MatType Type of matrix used.
+ */
+template<typename MatType>
+class AddReductionType
+{
+ public:
+  /**
+   * Reduce the specified network and store the results in the given
+   * parameter.
+   *
+   * @param layerOutputs Network output that should be reduced.
+   * @param output The network output.
+   */
+  void Reduce(const std::vector<MatType>& layerOutputs,
+              MatType& output)
+  {
+    output.zeros();
+		for (size_t i = 0; i < layerOutputs.size(); i++)
+		{
+			output += layerOutputs[i];
+		}
+  }
+
+	/**
+   * Reduce the specified network and store the results in the given
+   * parameter.
+   *
+   * @param input Network layer wise delta output.
+   * @param networkSize The network size.
+	 * @param layerDeltas The network deltas output.
+   */
+	void UnReduce(
```

Comment:
At least for simple reductions like add or product, couldn't you always just pass `input` to the `Backward()` call of each of the reduced layers?

Reply:
Yes, but for concat I can't do that.

Reply:
I'll open the PR for `Concat` shortly (I hit a few more bugs than I thought I would...) but maybe it is better to keep the two separate.  For `Concat` there is also the functionality that you can concatenate along any axis.  I think the implementation here is limited to the last axis.  Or are you planning on adding axis-specific support here too?

In any case, I'll open the PR for `Concat` later today (whether or not it's all debugged :smile: ) and maybe you can take some inspiration from there.  If you want this to do the same functionality `Concat` does that's just fine and we can remove `Concat` and use this instead, I'm not picky. :)